seed-data-scripts/replace-doi-urls/script.js

```diff
@@ -0,0 +1,149 @@
+      },
+      ...
+ * ]
+ *
+ * The result will be written in a new [seed_data]_updated.txt file
+ *
+ * To run the script:
+ * - run the script with the following command: 'node script.js [seed_data].txt [references_file].json'
+ */
+
+const fs = require("fs");
+
+// Make sure we got the filenames on the command line.
+if (process.argv.length < 4) {
+  console.log(
+    "PLEASE PROVIDE THE SEED DATA FILE NAME AND THE REFERENCES JSON FILE NAME!"
+  );
+  process.exit(1);
+}
+
+const SEED_DATA_FILE = process.argv[2];
+const REFERENCES_JSON = process.argv[3];
+
+if (SEED_DATA_FILE.split(".").pop() !== "txt") {
+  console.log("PLEASE PROVIDE A VALID SEED DATA TXT FILE!");
+  process.exit(1);
+}
+
+if (REFERENCES_JSON.split(".").pop() !== "json") {
+  console.log("PLEASE PROVIDE A VALID REFERENCES JSON FILE!");
+  process.exit(1);
+}
+
+const referencesArray = require(`./${REFERENCES_JSON}`);
+
+if (!Array.isArray(referencesArray)) {
+  console.log(
+    "REFERENCES JSON FILE MUST CONTAIN AN ARRAY WITH REFERENCES OBJECT!"
+  );
+  process.exit(1);
+}
+
+const getURLsFromText = (text) =>
+  text.match(
+    /(http|ftp|https):\/\/([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:()<>;\/~+#-]*[\w@?^=%&\/~+#-()])/g
+  ) || [];
+
+const validateDOI = (value) =>
+  /(10\.[0-9a-zA-Z]+\/(?:(?!["&\'])\S)+)\b/.test(value);
+
+const replaceDOIs = () => {
+  const newFileData = [];
+
+  try {
+    fs.readFile(SEED_DATA_FILE, "utf8", (error, data) => {
+      if (error) throw error;
+
+      const fileData = data.split("\n");
+
+      for (let i = 0; i < fileData.length; i++) {
+        let currentLine = fileData[i];
+
+        // extract doi urls from the current line
+        const doiUrls = [
+          ...new Set(
+            getURLsFromText(currentLine).filter((url) => validateDOI(url))
+          ),
+        ];
+
+        if (!doiUrls.length) {
+          // add text line as it is
+          newFileData.push(currentLine);
+          continue;
+        }
+
+        doiUrls.forEach((doiUrl) => {
+          // take first part of the text
+          const firstPart = currentLine.slice(0, currentLine.indexOf(doiUrl));
+
+          // take second part of the text and remove all occurrences of this DOI url
+          const secondPart = currentLine
+            .slice(currentLine.indexOf(doiUrl))
+            .split(doiUrl)
+            .filter((part) => part !== "\t")
+            .join("");
+
+          // find corresponding reference text for this DOI url
+          const referenceText = referencesArray
+            .find(
+              (reference) =>
+                reference[doiUrl] &&
+                typeof reference[doiUrl] === "string" &&
+                !reference[doiUrl].includes("<!DOCTYPE") // ignore html pages
```

Comment:
we may need to broaden this to anything we don't like e.g if it matches

(&[a-zA-Z]*;)|[<]

until we decide how to handle those in OCT-460